### DISCUSSION TITLE
Zustand is not updating nested object based state

### DISCUSSION THREAD

**Discussion by crisfcodes** [pmndrs/zustand#986] [Q&A]
I'm experimenting a problem with Zustand, I'm updating a state that is based on another state that is in a nested object, this state is not being updated properly... here an example:

```typescript
mapUpdate: (update: Partial<MapState['map']>) => {
  set((state) => {
    return {
      ...state,
      map: {
        ...state.map,
        ...update,
        canShowCarouselResults:
          state.map.zoom >= MIN_MAP_RESULTS_CAROUSEL_ZOOM,
      },
      canShowMarketResults: state.map.zoom <= MIN_RESULTS_ZOOM,
    };
  });
}
```

I need to update `canShowMArketResults` according to the zoom state, zoom state is in the map state object... the zoom state is updating correctly but the `canShowMarketResoults` status is not taking some map. zoom changes.

**Comment by dai-shi** [comment]
There are various approaches to this problem, but the first option is to not put derived state in store. You want to use selector to derive state.

```js
const canShowMarketResults = useStore((state) => state.map.zoom <= MIN_RESULTS_ZOOM)
```

**Reply by cflorez10** [reply]
@dai-shi thank for answering, so I need to do this in every component that needs canShowMarketResults? there is no other way?

**Reply by dai-shi** [reply]
You can create a custom hook, right?
```js
const useCanShowMarketResults = () => {
  return useStore((state) => state.map.zoom <= MIN_RESULTS_ZOOM)
}
```

**Reply by cflorez10** [reply]
is there any way to handle this directly in the store? is it an anti-pattern for zustand?

**Reply by dai-shi** [reply]
As I said, there are some approaches. If using selector is not an option, I would make actions to update derived properties at once. For example:

```js
const useStore = create((set) => ({
  count: 0,
  isEven: true,
  inc: () => set((state) = > {
    const count = state.count + 1
    return { count, isEven: (count % 2) === 0 }
  }),
  dec: () => set((state) = > {
    const count = state.count - 1
    return { count, isEven: (count % 2) === 0 }
  }),
}))
```

Another option would be to use 3rd party packages such as: https://github.com/cmlarsen/zustand-middleware-computed-state
